chore: Add a Jupyter style guide and supporting infra

thatlittleboy opened the pull request:
## Overview

This PR is in support and born out of PR #3037. The idea is to use this thread to discuss how we should handle Jupyter notebooks so that we can receive better quality contributions for #3036.

Many of the changes here are inspired by https://www.pymc.io/projects/docs/en/stable/contributing/jupyter_style.html.

### Discussions w.r.t. style guide

1. Using sphinx cross-references rather than URLs to refer to docstrings or other notebooks  \[**Status**: Under discussion\]
1. Add pre-commit to enforce notebook styles  \[**Status**: Active\]
    - `ruff` linting in code cells via `ruff`
    - `black` formatting in code cells via `black-jupyter` (instead of `nbqa` so that when we eventually apply black to the code-base, we can automatically ensure the black version applied to both `.py` and `.ipynb` is the same).
    - pygrep check for `watermark` extension (inspired from pymc docs)
1. Add [watermark](https://github.com/rasbt/watermark) cell to the notebooks. (enforced by pre-commit)
2. <s>Using `nb-clean` to clear notebook metadata (TBD)</s> We'll come back to this again if needed.


### Outside the scope of this PR

I propose the following scopes to be **outside** of the initial style guide + the efforts over at #3036 (at least, initially). These should probably be their own issues.

- [ ] Integrating MyST: I propose that this should be a separate effort on its own, and not be lumped together with #3036. It's a pretty large structural change on its own, and detracts from the main objective of #3036, which is to ensure the notebooks are runnable and updated. That said, I do like MyST and would be happy to transition to MyST in the medium term.
- [ ] Executing notebooks in CI/CD rather than committing in the notebook outputs. The idea of this is to reduce merge conflicts and diffs by only committing in the notebook source. Then the notebooks are executed in CI/CD in a controlled environment, before we run sphinx. This makes sure that the notebooks are always up-to-date, and errors if we change the API but forget to change the notebook examples. Again, this is nice-to-have, but I think is outside the scope.


## Checklist

- [X] Supports #3036.  <!--Replace xxxx with the GitHub issue number-->
- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [X] Unit tests added (if fixing a bug or adding a new feature)
- [ ] Added entry to `CHANGELOG.md` (if changes will affect users)

thatlittleboy committed:
add watermark to docs requirements

thatlittleboy committed:
add precommit checks

thatlittleboy committed:
use autosectionlabel to generate targets automatically

thatlittleboy committed:
add style guide to CONTRIBUTING

connortann commented:
There seems to be a bit of cleaning up we can do: there are plenty of notebooks in HTML format in `./docs/notebooks`, which I don't _think_ are used anywhere. It would be useful to understand the history and whether they need to be kept.

connortann commented:
Another suggestion: the checklist for docs (which could perhaps be mentioned on the meta-issue) could include:

"Ensure that the notebook does not raise syntax warnings in the Sphinx build logs"

E.g., this kind of thing:

```
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/docs/example_notebooks/tabular_examples/linear_models/Explaining a model that uses standardized features.ipynb:3: WARNING: Each notebook should have at least one section title
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/docs/example_notebooks/tabular_examples/linear_models/Math behind LinearExplainer with correlation feature perturbation.ipynb:54: ERROR: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/docs/example_notebooks/tabular_examples/linear_models/Math behind LinearExplainer with correlation feature perturbation.ipynb:52: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/docs/example_notebooks/tabular_examples/linear_models/Math behind LinearExplainer with correlation feature perturbation.ipynb:55: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/shap/explainers/_deep/__init__.py:docstring of shap.explainers._deep.Deep.__init__:14: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/shap/explainers/_deep/__init__.py:docstring of shap.explainers._deep.Deep.__init__:19: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/shap/maskers/_image.py:docstring of shap.maskers._image.Image.__init__:6: ERROR: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/shap/checkouts/3110/shap/maskers/_image.py:docstring of shap.maskers._image.Image.__init__:7: WARNING: Block quote ends without a blank line; unexpected unindent.
```

connortann commented:
@thatlittleboy  it would be great to resolve the conflicts on this PR and get it merged in. I think publishing this guidance is helpful for growing our contributor community, as fixing the notebooks is a great `good-first-issue`.

thatlittleboy commented:
@connortann Got it, I'll work on this this week.

thatlittleboy commented:
Pre-commit checks are failing because of the watermarks check. I'll fix these all at once after we merge in the remaining notebooks PR in our backlog, but otherwise this PR is ready to receive comments & review.

@connortann @dsgibbons

connortann commented on CONTRIBUTING.md:
```diff
@@ -114,6 +115,9 @@ Consequently, is is quite normal to see warnings such as `WARNING: Could not
 available.
 
+Install the `docs` extras as well, if you are modifying the documentation / example
+notebooks.
```
I think this comment would be better placed around line 108, describing the `docs` extra. It is already included as an extra to be installed on line 99.

thatlittleboy replied:
Yea good call, this was added before the semi-recent changes that you made. I'll remove these line then

connortann commented on CONTRIBUTING.md:
```diff
@@ -207,9 +211,74 @@ To build the documentation locally:
 1. Navigate to the `docs` directory.
 2. Run `make html`.
-3. Open "_build/html/index.html" in your browser.
+3. Open "\_build/html/index.html" in your browser to inspect the documentation.
```
Is this backslash at the start of the link intentional?

thatlittleboy replied:
Hmm it was added in by my autoformatter automatically, to escape the underscore. I think it renders fine on GH

![image](https://github.com/shap/shap/assets/30731072/c903b91f-f348-4132-b5e2-cacd3b795042)

connortann commented on CONTRIBUTING.md:
```diff
@@ -207,9 +211,74 @@ To build the documentation locally:
 [nbsphinx installation
 guide](https://nbsphinx.readthedocs.io/en/0.9.2/installation.html#pandoc).
+
+### Jupyter notebook style guide
```
Nitpick: I would suggest bumping the heading level of this section up one level, i.e.:

```
## Jupyter notebook style guide
### General jupyter guidelines
et cetera
```

The reason being, we might wish to put these contributing guidelines on Sphinx at some point in future, and with the ReadTheDocs theme it's generally helpful for the layout if there are only two main levels of headings used throughout each document, rather than 3 or more levels of nesting.

connortann commented on CONTRIBUTING.md:
```diff
@@ -207,9 +211,74 @@ To build the documentation locally:
+instead.
+
+We have `sphinx.ext.autosectionlabel` extension enabled, which autogenerates targets
+for every section, named `{relative/path/to/page}:{title-of-section}`.
```
It would be helpful here to give an example of a valid cross reference to a notebook.

thatlittleboy replied:
@connortann I couldn't get the cross-reference working using the sphinx targets.

I'm following the `nbsphinx` documentation instead, which recommends [this](https://nbsphinx.readthedocs.io/en/0.9.3/markdown-cells.html#Links-to-Domain-Objects).

It's not as elegant, but it works for now. I recently used it in PR #3336, see the output notebook [here](https://shap.readthedocs.io/en/latest/example_notebooks/tabular_examples/tree_based_models/Force%20Plot%20Colors.html): 

![image](https://github.com/shap/shap/assets/30731072/b0ee02f5-4c5d-4604-998d-d28979cb83c0)

connortann replied:
I would be good to get usual cross-references working; I'll see if I can take a look later.

Let's not hold up this PR for that though, as this is a good starting point!

connortann commented on .pre-commit-config.yaml:
```diff
@@ -38,10 +38,25 @@ repos:
     types_or: [jupyter]
     files: *cleanNotebooks
+
+- repo: local
```
I'm in two minds on whether it's worth implementing the watermark as a pre-commit hook for notebooks. Pre-commit hooks are certainly nice for linting, but this might be a bit to strict.

Watermarks make sense for fairly long examples, especially those that use heavy 3rd party libraries, take a while to execute, or might be hard to reproduce / be slow running. However, for short API demos or introductory pages I think the watermark doesn't add much value, and is not worth the added visual clutter. So, not sure it's wise to add it as an absolutely mandatory hook for all notebooks.

In any case, I would suggest removing or commenting this hook out for now, so we can merge the PR and get the contributing guidelines published ASAP.

connortann reviewed (approved):
Pre-approving with some suggestions above

thatlittleboy committed:
refactor: code review

thatlittleboy committed:
remove watermark

connortann commented:
@thatlittleboy over to you to merge when happy!

thatlittleboy committed:
refactor: adjust prose about cross-references in notebooks

Couldn't get the cross references to work, I'm following this
'https://nbsphinx.readthedocs.io/en/0.9.3/markdown-cells.html#Links-to-Domain-Objects'
instead

thatlittleboy commented:
@connortann I revised the section about the cross-references slightly, see my reply to your review comment. Let me know if you have any other comments, thanks!
